dips/dip-213.md

```diff
@@ -0,0 +1,94 @@
+but makes bottlenecks more visible and could guide future optimizations.
+
+# Description
+
+In [DiemBFTv4](https://developers.diem.com/papers/diem-consensus-state-machine-replication-in-the-diem-blockchain/2021-08-17.pdf), a block (proposal) goes through different steps before it's finalized. The steps are 
+```
+Consensus stage
+    1. Proposed
+    2. Executed
+    3. Voted (including execution result)
+    4. QuorumCertified
+    5. 2-chain Certified (Finalized)
+```
+
+The proposed change is to decouple execution from consensus and pipeline the process with different stages. A block would go through
+4 stages each with its own steps:
+```
+1. Consensus stage 
+    * Proposed
+    * Voted (no execution result)
+    * QuorumCertified
+    * 2-chain Certified (Ordered)
+2. Execution stage
+3. Voting stage (for execution result)
+4. Commit stage (Finalized)
+```
+The stages run in parallel, utilizing available compute resources to increase overall throughput. 
+That is, the system could commit B1, sign B2, execute B3, and order B4, all at the same time. Note, while throughput increases due to parallel execution, latency of an individual transaction (block) finalization does not decrease, as the stages are inherently sequential. In fact, individual transaction latency may slightly increase since commit occurs less "eagerly".
+
+# Required Changes
+
+## Consensus
+`StateComputer` is the trait that consensus uses for `compute`, `commit` or `sync_to` blocks.
+An `ordering_state_computer` is implemented to bypass execution and send blocks to next stage when ordered.
+
+## BufferManager
+A `BufferManager` is implemented to manage different stages of ordered blocks, it's a queue of blocks with stage markers.
+- `Ordered` block is sent to execution and advanced to `Executed` after receiving execution result.
+- `Executed` block is sent to safety rules and advanced to `Signed` after receiving a signature.
+- Signature on `Signed` block is broadcasted to all validators. At all stages except `Aggregated`, signatures from other validators on the block are collected.
+- A block in either `Executed` or `Signed` stage is advanced to `Aggregated` after receiving a quorum of signatures.
+- `Aggregated` block is popped from the queue and sent to persistent storage.
+ 
+## Sync
+When the node is far from current proposed block, it may decide to fast-forward via state sync protocol. It's triggered if the gap between
+local highest round of ordered block and remote highest round of committed block becomes large (current threshold is 10). Upon state sync, 
+the node fast-forward to the same state (both ledger state and pending blocks) as the remote peer. 
+
+## Backpressure
+To prevent consensus-ordering going too fast and create ever-growing backlog for other stages, back pressure is implemented to stop ordering from making progress if the gap between the highest committed block and highest ordered block is large.
+
+## Reconfiguration
```

Comment:
please explain: why are we describing this here, and what is the connection to decoupled execution?

Reply:
before decoupled-execution, proposer is aware of reconfiguration and propose empty suffix (also enforced by voters).
after the change, each node decides to skip the suffix during execution (ordering is not aware of reconfiguration anymore).

Reply:
First, it's definitely worth explaining this carefully here.

Second, this concerns me a bit: Who and when creates a new epoch genesis then?

Reply:
the (virtual) genesis block is derived from the ledger info from the previous epoch. the deriving function remains unchanged.
with block batches [1 -> 2] -> [r -> 3] -> [4] -> [5, 6], the ledger info would be aggregated on 3 (txns in 3 are skipped).

Reply:
Got it, but all worth elaborating here -- 4, 5, 6 -- skipped? when does a leader propose in the new epoch? etc etc

Reply:
yes, 4, 5, 6 are all skipped, nodes block on aggregating ledger info on 3 (periodically broadcast signature until commit).  I was trying to only discuss the delta here though - how is genesis derived, when does leader enter new epoch, how do nodes catch up etc remains unchanged

Reply:
I don't think it's the same nor trivial. 

IIUC, a leader waits for commit votes on 3 and then discards all blocks after 3, forms an epoch genesis, and proposes it, right? Let's say a few more words to this effect ;-)